gunrock/app/sm/sm_problem.cuh

```diff
@@ -460,9 +475,18 @@ struct Problem : ProblemBase<_GraphT, _FLAG> {
         cout << endl;
       }
       cout << endl;*/
-      h_subgraphs[0] = combinations.size();
-      // TODO: export combinations to output
-
+      // returning results will be stored on the CPU
+      if (device == "CPU") {
+        count_subgraphs[0] = combinations.size();
+        list_subgraphs = new VertexT[combinations.size()];
+        std::copy(combinations.begin(), combinations.end(), list_subgraphs);
+      } else { // returning results will be stored on the GPU
+        VertexT *h_subgraphs = new VertexT[1];
+        h_subgraphs[0] = combinations.size();
+        GUARD_CU(cudaMemcpy(count_subgraphs, h_subgraphs, 1, cudaMemcpyHostToDevice));
+        GUARD_CU(cudaMalloc(list_subgraphs, combinations.side()));
```

Comment:
So this suggests that the internal algorithm is generating the results in host memory and you're going to copy it back out into device memory?  While that will achieve my goal, wouldn't it be more efficient overall if the results were generated directly into GPU memory?

I haven't worked through the logic in the implementation, but at some point there is a CUDA kernel that is iterating over the CSR representation of the graph and generating a sequence of subgraph matches.  If you're storing them in host memory, aren't you paying a transfer cost (which we typically assume is expensive) to bring those subgraphs out of the CUDA kernel into host memory... and now we'll pay that cost again to put it back into device memory.

It would seem that generating the list of subgraph matches directly in device memory would be more efficient.

Reply:
Yes, you are correct. We should in fact just return the subgraphs that already exist on the device and are accessible in the problem::extract().